src/box/box.cc

```diff
@@ -1805,6 +1805,220 @@ box_wait_quorum(uint32_t lead_id, int64_t target_lsn, int quorum,
 	return 0;
 }
 
+/** A structure holding trigger data to collect syncs. */
+struct sync_trigger_data {
+	/** Syncs to wait for. */
+	uint64_t syncs[VCLOCK_MAX];
+	/**
+	 * A bitmap holding replica ids whose vclocks were already collected.
+	 */
+	vclock_map_t collected_vclock_map;
+	/** The fiber waiting for vclock. */
+	struct fiber *waiter;
+	/** Collected vclock. */
+	struct vclock *vclock;
+	/** The request deadline. */
+	double deadline;
+	/** How many vclocks are needed. */
+	int count;
+	/** Whether the request is timed out. */
+	bool is_timed_out;
+};
+
+/**
+ * A trigger executed on each ack to collect up to date remote node vclocks.
+ * When an ack comes with requested sync for some replica id, ack vclock is
+ * accounted.
+ */
+static int
+check_sync_on_ack(struct trigger *trigger, void *event)
+{
+	struct replication_ack *ack = (struct replication_ack *)event;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = ack->source;
+	uint64_t sync = data->syncs[id];
+	int accounted_count = bit_count_u32(data->collected_vclock_map) + 1;
+	if (!bit_test(&data->collected_vclock_map, id) && sync > 0 &&
+	    ack->sync >= sync && accounted_count < data->count) {
+		vclock_max_ignore0(data->vclock, ack->vclock);
+		bit_set(&data->collected_vclock_map, id);
+		++accounted_count;
+		if (accounted_count >= data->count)
+			fiber_wakeup(data->waiter);
+	}
+	return 0;
+}
+
+/** A trigger querying relay's next sync value once it becomes operational. */
+static int
+relay_get_sync_on_start(struct trigger *trigger, void *event)
+{
+	struct replica *replica = (struct replica *)event;
+	if (replica->anon)
+		return 0;
+	struct relay *relay = replica->relay;
+	struct sync_trigger_data *data =
+		(struct sync_trigger_data *)trigger->data;
+	uint32_t id = replica->id;
+	/* Already accounted. */
+	if (bit_test(&data->collected_vclock_map, id))
+		return 0;
+	if (relay_next_sync(relay, &data->syncs[id], data->deadline) != 0) {
```

Comment:
I added an `assert(false)` here and the tests passed. Maybe worth adding a test, at least a small simple one?

Reply:
I added a test: seems `relay_get_sync_on_start` wasn't tested at all.
I only tested the success route though, because making `relay_next_sync` timeout in `relay_get_sync_on_start` before the waiting times out in `box_collect_confirmed_vclock` is tricky.
<details><summary>diff</summary>

```diff
diff --git a/test/replication-luatest/linearizable_read_test.lua b/test/replication-luatest/linearizable_read_test.lua
index d6ef275e4..33c48a4ce 100644
--- a/test/replication-luatest/linearizable_read_test.lua
+++ b/test/replication-luatest/linearizable_read_test.lua
@@ -1,9 +1,10 @@
 local t = require('luatest')
 local cluster = require('test.luatest_helpers.cluster')
 local server = require('test.luatest_helpers.server')
+local proxy = require('test.luatest_helpers.proxy.proxy')
 local g = t.group('linearizable-read')
 
-local proxy = require('test.luatest_helpers.proxy.proxy')
+local fiber = require('fiber')
 
 local function build_replication(num_instances)
     local t = {}
@@ -18,7 +19,7 @@ local num_servers = 3
 g.before_all(function()
     g.cluster = cluster:new({})
     g.servers = {}
-    local box_cfg = {
+    g.box_cfg = {
         replication = build_replication(num_servers),
         replication_timeout = 0.1,
         txn_isolation = 'read-confirmed',
@@ -26,16 +27,16 @@ g.before_all(function()
     }
     g.servers[1] = g.cluster:build_and_add_server({
         alias = 'server_1',
-        box_cfg = box_cfg,
+        box_cfg = g.box_cfg,
     })
     -- Servers 2 and 3 are interconnected without a proxy.
     for i = 2, num_servers do
-        box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
+        g.box_cfg.replication[i] = server.build_instance_uri('server_' .. i)
     end
     for i = 2, num_servers do
         g.servers[i] = g.cluster:build_and_add_server({
             alias = 'server_' .. i,
-            box_cfg = box_cfg,
+            box_cfg = g.box_cfg,
         })
     end
     g.proxies = {}
@@ -148,6 +149,32 @@ g.test_no_dirty_reads = function(g)
     end)
 end
 
+g.test_reconnect_while_waiting = function(g)
+    for i = 2, 3 do
+        g.servers[i]:exec(function()
+            local repl = table.copy(box.cfg.replication)
+            table.remove(repl, 1)
+            box.cfg{replication = repl}
+        end)
+    end
+    local f = fiber.new(g.servers[1].exec, g.servers[1], function()
+        local ok, err = pcall(box.begin, {linearizable = true})
+        if ok then
+            box.commit()
+        end
+        return ok, err
+    end)
+    f:set_joinable(true)
+    for i = 2, 3 do
+        g.servers[i]:exec(function(repl)
+            box.cfg{replication = repl}
+        end, {g.box_cfg.replication})
+    end
+    local success, ok, err = f:join()
+    t.assert(success, 'fiber joined successfully')
+    t.assert(ok, 'reconnect while begin() waits for vclocks is tolerated')
+end
+
 g.test_leader_change = function(g)
     g.servers[1]:exec(function()
         box.ctl.promote()
```
</details>